dbms/src/Interpreters/JoinV2/HashJoinProbe.cpp

```diff
@@ -385,40 +800,58 @@ void NO_INLINE JoinProbeBlockHelper<KeyGetter, has_null_map, tagged_pointer>::jo
         if (state->stage == ProbePrefetchStage::FindNext)
         {
             RowPtr ptr = state->ptr;
-            RowPtr next_ptr = HashJoinRowLayout::getNextRowPtr(ptr);
-            if (next_ptr)
-            {
-                state->ptr = next_ptr;
-                PREFETCH_READ(next_ptr);
-            }
+            RowPtr next_ptr = getNextRowPtr(ptr);
+            state->ptr = next_ptr;
 
             const auto & key2 = key_getter.deserializeJoinKey(ptr + key_offset);
             bool key_is_equal = joinKeyIsEqual(key_getter, state->key, key2, state->hash, ptr);
             collision += !key_is_equal;
             if (key_is_equal)
             {
-                ++current_offset;
-                selective_offsets.push_back(state->index);
-                insertRowToBatch(key_getter, ptr + key_offset, key2);
-                if unlikely (current_offset >= settings.max_block_size)
+                if constexpr (Adder::need_not_matched)
+                    state->is_matched = true;
+
+                if constexpr (Adder::need_matched)
                 {
-                    if (!next_ptr)
+                    bool is_end = Adder::addMatched(
+                        *this,
+                        wd,
+                        added_columns,
+                        state->index,
+                        current_offset,
+                        ptr,
+                        key_offset + key_getter.getRequiredKeyOffset(key2));
+                    if unlikely (is_end)
                     {
-                        state->stage = ProbePrefetchStage::None;
-                        --active_states;
+                        if constexpr (Adder::break_on_first_match)
+                            next_ptr = nullptr;
+
+                        if (!next_ptr)
+                        {
+                            state->stage = ProbePrefetchStage::None;
+                            --active_states;
+                        }
+                        break;
                     }
-                    break;
+                }
+
+                if constexpr (Adder::break_on_first_match)
+                {
+                    next_ptr = nullptr;
```

Comment:
next_ptr is set to nullptr, but `state->ptr` still hold the original value of `next_ptr`, is it expected?

Reply:
Yes. This `state` stage will be `None` after `next_ptr = nullptr`, so it does not matter whether `state->ptr` is valid or not.